timemachine/fe/free_energy.py

```diff
@@ -1120,6 +1136,91 @@ def compute_dense():
     return U_kl
 
 
+def make_u_kl_fxn(trajs, initial_states):
+    """fxn(k, l) = "trajs[k] evaluated in ensembles[l]"
+
+    usage note: be careful of axis-ordering convention, see: https://github.com/proteneer/timemachine/issues/1100
+    """
+
+    summed_potentials = [make_summed_potential(s.potentials) for s in initial_states]
+
+    # validate assumption that initial states all have compatible potentials / ensembles
+    kBTs = [BOLTZ * state.integrator.temperature for state in initial_states]
+    s_0 = initial_states[0]
+    for s in initial_states[1:]:
+        assert_ensembles_compatible(s_0, s)
+        assert_potentials_compatible(s_0.potentials, s.potentials)
+
+    sp = summed_potentials[0]
+    sp_gpu = sp.potential.to_gpu(np.float32)
+
+    def batch_U_fxn(xs, ps, bs, x_idxs, p_idxs):
+        Us = sp_gpu.unbound_impl.execute_batch_sparse(xs, ps, bs, x_idxs, p_idxs, False, False, True)[2]
+        return np.nan_to_num(Us, nan=+np.inf)
+
+    def u_kl(k, l):
+        coords = trajs[k].frames
+        boxes = trajs[k].boxes
+
+        params = np.array([summed_potentials[l].params])
+
+        coords_batch_idxs = np.arange(len(coords)).astype(np.uint32)
+        params_batch_idxs = np.zeros_like(coords_batch_idxs).astype(np.uint32)
+
+        Us = batch_U_fxn(coords, params, boxes, coords_batch_idxs, params_batch_idxs)
+
+        return Us / kBTs[l]
+
+    return u_kl
+
+
+def assert_ensembles_compatible(state_a: InitialState, state_b: InitialState):
+    """check that xvb from state_a can be swapped with xvb from state_b (up to timestep error),
+    with swap acceptance probability that depends only on U_a, U_b, kBT"""
+
+    # assert (A, B) have identical masses, temperature
+    intg_a = state_a.integrator
+    intg_b = state_b.integrator
+
+    assert (intg_a.masses == intg_b.masses).all()
+    assert intg_a.temperature == intg_b.temperature
+
+    # assert same pressure (or same volume)
+    assert (state_a.barostat is None) == (state_b.barostat is None), "should both be NVT or both be NPT"
+
+    if state_a.barostat and state_b.barostat:
+        # assert (A, B) are compatible NPT ensembles
+        baro_a: MonteCarloBarostat = state_a.barostat
+        baro_b: MonteCarloBarostat = state_b.barostat
+
+        assert baro_a.pressure == baro_b.pressure
+        assert baro_a.temperature == baro_b.temperature
+
+        # also, assert barostat and integrator are self-consistent
+        assert intg_a.temperature == baro_a.temperature
+
+    else:
+        # assert (A, B) are compatible NVT ensembles
+        assert (state_a.box0 == state_b.box0).all()
+
+
+def compute_u_kn(trajs, initial_states) -> Tuple[NDArray, NDArray]:
+    """makes K^2 calls to execute_batch_selective"""
```

Comment:
nit: should this be
```suggestion
    """makes K^2 calls to execute_batch_sparse"""
```
?

(where each call to `execute_batch_sparse` is doing `trajs.shape[1]` calls to `execute_batch_selective`?)

Reply:
Good catch -- fixed in https://github.com/proteneer/timemachine/pull/1387/commits/3523176d4efa5ed88ca1691160a437182bd28941